runtime/argument_validation.go

```diff
@@ -0,0 +1,354 @@
+/*
+ * Cadence - The resource-oriented smart contract programming language
+ *
+ * Copyright 2019-2021 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package runtime
+
+import (
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/interpreter"
+	"github.com/onflow/cadence/runtime/sema"
+)
+
+// Utility functions for validating the arguments passed into a script or a transaction.
+
+// valueConformsToDynamicType checks whether the shape of the value is same as the shape
+// defined by the type associated with the value. Returns `true` if the value and the
+// type matches, and `false` otherwise.
+//
+// e.g: A value with the type as 'Foo', but having additional fields than what is
+// specified in 'Foo', would return `false`.
+//
+func valueConformsToDynamicType(value interpreter.Value, dynamicType interpreter.DynamicType) (ok bool) {
+
+	// Currently the type-vs-value mismatch could only happen for composite types, as
+	// that's the only place where the type is loaded from the runtime, based on the name,
+	// rather than based on the value structure of the user input.
+	// Hence the possible set of malformed values are:
+	//     - Composite types.
+	//     - Arrays/Dictionaries/Optionals of composite types.
+
+	switch typ := dynamicType.(type) {
```

Comment:
Maybe instead of having this large type switch and we forget to update it when adding more types in the future, we could add a function to the `interpreter.DynamicType` interface, e.g. `ValueConforms(value Value) bool`, just like / analogous to the `interpreter.Value.DynamicType()` function which returns the dynamic type for a value?

Reply:
Yeah, that's a good idea. Initially, I didn't want to pollute the existing interface and the structs. But yeah, it is easy to miss updating this switch-case.